sdk/core/azure-core/src/main/java/com/azure/core/http/policy/HttpLoggingPolicy.java

```diff
@@ -117,44 +117,59 @@ private HttpLoggingPolicy(HttpLogOptions httpLogOptions, boolean prettyPrintJson
 
         addHeadersToLogMessage(request.getHeaders(), requestLogMessage, numericLogLevel);
 
-        Mono<String> requestLoggingMono = Mono.defer(() -> Mono.just(requestLogMessage.toString()));
+        if (!httpLogDetailLevel.shouldLogBody()) {
+            return logAndReturn(logger, requestLogMessage, null);
+        }
 
-        if (httpLogDetailLevel.shouldLogBody()) {
-            if (request.getBody() == null) {
-                requestLogMessage.append("(empty body)")
-                    .append(System.lineSeparator())
-                    .append("--> END ")
-                    .append(request.getHttpMethod())
-                    .append(System.lineSeparator());
-            } else {
-                String contentType = request.getHeaders().getValue("Content-Type");
-                long contentLength = getContentLength(logger, request.getHeaders());
+        if (request.getBody() == null) {
+            requestLogMessage.append("(empty body)")
+                .append(System.lineSeparator())
+                .append("--> END ")
+                .append(request.getHttpMethod())
+                .append(System.lineSeparator());
 
-                if (shouldBodyBeLogged(contentType, contentLength)) {
-                    requestLoggingMono = FluxUtil.collectBytesInByteBufferStream(request.getBody()).flatMap(bytes -> {
+            return logAndReturn(logger, requestLogMessage, null);
+        }
+
+        String contentType = request.getHeaders().getValue("Content-Type");
+        long contentLength = getContentLength(logger, request.getHeaders());
+
+        if (shouldBodyBeLogged(contentType, contentLength)) {
+            ByteArrayOutputStream outputStream = new ByteArrayOutputStream((int) contentLength);
+
+            // Add non-mutating operators to the data stream.
+            request.setBody(
+                request.getBody()
+                    .doOnNext(byteBuffer -> {
+                        for (int i = byteBuffer.position(); i < byteBuffer.limit(); i++) {
```

Comment:
I am not sure how expensive this operation is going to be. Would iterating through the bytebuffer slow down the application just for logging? Would it be good to also have support for truncated logs for body - for e.g. if a flag is set to log only first few bytes of the body just to show what the payload's looking like but not the entire body.

Reply:
I don't think there should be a large performance issue by iterating over the `ByteBuffer` using `get(int index)` given how the bulk `get(byte[] dst, int offset, int length)` method works internally. The bulk get method uses `get()` internally and has the side effect of mutating the `ByteBuffer`'s position. In terms of getting a smaller portion of the body I think that could make sense, right now body logging hard caps at 16KB and anything larger the body is ignored during logging, we could instead shrink that to 1KB or 4KB (just something small) and always log it.

Reply:
https://github.com/Azure/azure-sdk-for-java/issues/6670